[COT] `query()` method: add support for field queries

jorgeatorres opened the pull request:
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR adds support for a `field_query` arg to `OrdersTableQuery`, which can be used via `wc_get_orders()` when COT is active.
This came to be as a result of an internal discussion (p1660893943441389-slack-C9QDV4TL4) where we wanted to add a way to easily query orders matching various fields and value combinations without the need to write custom SQL queries. This wasn't really possible before, but thanks to most of our fields being metadata, `WP_Query`'s `meta_query` could be used for that purpose. As such, `field_query` supports most (or all?) of `meta_query`'s (metavalue) functionality but operates on core order fields instead of metadata.

With the new `field_query` you can perform complex queries: matching various fields and with AND/OR relations. For example, to obtain all pending/on-hold orders where the amount > 10.0 and the purchaser's first name or last name contains "michel" you could do this:

```php
$orders = wc_get_orders(
	[
		'field_query' => [
			[
				'field'   => 'total',
				'value'   => '10.0',
				'compare' => '>',
				'type'    => NUMERIC,
			],
			[
				'relation' => 'OR',
				[
					'field'   => 'billing_first_name',
					'value'   => 'michel',
					'compare' => 'LIKE',
				],
				[
					'field'   => 'billing_last_name',
					'value'   => 'michel',
					'compare' => 'LIKE',
				]
			]
		],
		'status' => [ 'wc-pending', 'wc-on-hold' ]
	]
);
```

The syntax for `field_query` clauses is analogous to that of WP's [meta query](https://developer.wordpress.org/reference/classes/wp_meta_query/) but with `field` standing for `key`.

Also, `field_query` can be combined with other types of queries already supported by `OrdersTableQuery` such as direct access to certain fields, `meta_query` and `date_query`.

### How to test the changes in this Pull Request:

#### Option 1
1. Check added unit tests.

#### Option 2
1. Activate and make COT authoritative.
2. Populate the db with some orders and data.
3. Run your own queries using `wc_get_orders()`. Join the fun.

### Next Steps
With `field_query` we can rewrite some of the code in `OrdersTableQuery`, dealing with basic columns in the core tables, in a more concise manner. My plan is to do that in a subsequent PR, as well adding an abstract class for this kind of "recursive" query to share code between `OrdersTableMetaQuery` and `OrdersTableFieldQuery`. We've also discussed the possibility of supporting these queries in the CPT implementation to make things simpler for devs.

Once this PR is reviewed/merged I'll submit PRs for the above.

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [X] Have you written new tests for your changes, as applicable?
-   [X] Have you successfully run tests with your changes locally?
-   [X] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

jorgeatorres committed:
Bring Yoda back.

jorgeatorres committed:
Add method to OrdersTableQuery to obtain table info for a given order field

jorgeatorres committed:
First pass at field_query

jorgeatorres committed:
Add tests

jorgeatorres committed:
Add changelog

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php:
```diff
@@ -0,0 +1,311 @@
+		$clause_compare = $clause['compare'];
+
+		switch ( $clause_compare ) {
+			case 'IN':
+			case 'NOT IN':
+				$where = $wpdb->prepare( '(' . substr( str_repeat( ',%s', count( $clause_value ) ), 1 ) . ')', $clause_value ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+				break;
+			case 'BETWEEN':
+			case 'NOT BETWEEN':
+				$where = $wpdb->prepare( '%s AND %s', $clause_value[0], $clause_value[1] );
+				break;
+			case 'LIKE':
+			case 'NOT LIKE':
+				$where = $wpdb->prepare( '%s', '%' . $wpdb->esc_like( $clause_value ) . '%' );
```
seems like `$clause_compare` is not SQL escaped for these cases, we should escape them.

jorgeatorres replied:
@vedanshujain: could you let me know which cases need to be SQL-escaped? I might be missing something, but as far as I can see, all are using `%s` placeholders.

vedanshujain replied:
You are right, they are indeed prepared. Not sure why I said that, thanks!

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php:
```diff
@@ -604,6 +619,55 @@ private function build_query(): void {
+		$join    = '';
+		$join_on = '';
+
+		$join .= "INNER JOIN {$table}" . ( $alias !== $table ? " AS {$alias}" : '' );
```
table name should be escaped in backticks

vedanshujain commented on plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php:
```diff
@@ -995,4 +995,104 @@ public function test_cot_query_search(): void {
+	/**
+	 * Ensure field_query works as expected.
+	 */
+	public function test_cot_query_field_query(): void {
```
Can we add more tests where
1. we do field along with other queries (meta, ids etc) at the same time
2. query result is empty
3. invalid value types
4. invalid column names

jorgeatorres replied:
I've added a few more tests, but please let me know if you want to see something else in particular.

vedanshujain reviewed (commented):
I'm still reviewing, but pretty neat work, as usual. I wonder, can we add a compatibility layer where we also read meta_queries, and if a meta_key is the same as a column name then we implicitly convert it into a field query? That way, existing meta_queries will work seamlessly on COT as well without needing any updation.

jorgeatorres committed:
Use backticks for table names and aliases

jorgeatorres committed:
Improve validation in field_query

jorgeatorres committed:
Add some more tests

jorgeatorres commented:
Hey @vedanshujain!

Thanks for the review. I've rebased to fix conflicts, and added some more validation and tests to the code. Please let me know what you think.

vedanshujain commented:
Looks great, thanks for your amazing work both here and with OrderTableQuery 🙌
